File: backend/app/api/routes/findings.py
import uuid

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_current_user
from app.db.session import get_db
from app.models.finding import Finding
from app.schemas.finding import FindingCreate, FindingOut, FindingUpdate

router = APIRouter(
    prefix="/api/engagements/{engagement_id}/findings",
    tags=["findings"],
    dependencies=[Depends(get_current_user)],
)


@router.get("", response_model=list[FindingOut])
async def list_findings(
    engagement_id: uuid.UUID, db: AsyncSession = Depends(get_db)
) -> list[Finding]:
    result = await db.execute(
        select(Finding)
        .where(Finding.engagement_id == engagement_id)
        .order_by(Finding.created_at.desc())
    )
    return list(result.scalars().all())


@router.post("", response_model=FindingOut, status_code=status.HTTP_201_CREATED)
async def create_finding(
    engagement_id: uuid.UUID, payload: FindingCreate, db: AsyncSession = Depends(get_db)
) -> Finding:
    finding = Finding(engagement_id=engagement_id, **payload.model_dump())
    db.add(finding)
    await db.commit()
    await db.refresh(finding)
    return finding


async def _get_finding_or_404(
    engagement_id: uuid.UUID, finding_id: uuid.UUID, db: AsyncSession
) -> Finding:
    finding = await db.get(Finding, finding_id)
    if finding is None or finding.engagement_id != engagement_id:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Hallazgo no encontrado")
    return finding


@router.patch("/{finding_id}", response_model=FindingOut)
async def update_finding(
    engagement_id: uuid.UUID,
    finding_id: uuid.UUID,
    payload: FindingUpdate,
    db: AsyncSession = Depends(get_db),
) -> Finding:
    finding = await _get_finding_or_404(engagement_id, finding_id, db)
    for field, value in payload.model_dump(exclude_unset=True).items():
        setattr(finding, field, value)
    await db.commit()
    await db.refresh(finding)
    return finding


@router.delete("/{finding_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_finding(
    engagement_id: uuid.UUID, finding_id: uuid.UUID, db: AsyncSession = Depends(get_db)
) -> None:
    finding = await _get_finding_or_404(engagement_id, finding_id, db)
    await db.delete(finding)
    await db.commit()

File: frontend/src/api/findings.ts
import { apiClient } from '@/lib/api'
import type { Finding, FindingInput } from './types'

export async function listFindings(engagementId: string): Promise<Finding[]> {
  const { data } = await apiClient.get<Finding[]>(`/api/engagements/${engagementId}/findings`)
  return data
}

export async function createFinding(
  engagementId: string,
  payload: FindingInput,
): Promise<Finding> {
  const { data } = await apiClient.post<Finding>(
    `/api/engagements/${engagementId}/findings`,
    payload,
  )
  return data
}

export async function updateFinding(
  engagementId: string,
  findingId: string,
  payload: Partial<FindingInput>,
): Promise<Finding> {
  const { data } = await apiClient.patch<Finding>(
    `/api/engagements/${engagementId}/findings/${findingId}`,
    payload,
  )
  return data
}

export async function deleteFinding(engagementId: string, findingId: string): Promise<void> {
  await apiClient.delete(`/api/engagements/${engagementId}/findings/${findingId}`)
}
